### ISSUE TITLE
CalculatePatchAndPushOnWebSocket gives wrong result when run syncronously

### ISSUE THREAD

**Issue by warpech** [Starcounter/Starcounter#4144]
Yesterday I have spent half of a day investigating a problem in SWO, which they have been investigting for even longer. It turns out that `Session.CalculatePatchAndPushOnWebSocket` gives strange kind of failures when used outside of `Session.ScheduleTask`.

The below snippet illustrates the problem:

```cs
using Starcounter;

namespace HelloWorld
{
    partial class SpenderPage : Json, IBound<Spender>
    {
        protected override void OnData()
        {
            //(1) unless you comment the below line, the app crashes (Session == null)
            Session.CalculatePatchAndPushOnWebSocket();
        }

        [SpenderPage_json.Spendings]
        partial class SpendingsItem : Json, IBound<Expense>
        {
            protected override void OnData()
            {
                //(2) unless you comment the below line, SpenderPage response is 
                //NOT wrapped in a namespace
                Session.CalculatePatchAndPushOnWebSocket();
            }
        }
    }   
}
```

The above snippet shows the two problems:

1. `Session.CalculatePatchAndPushOnWebSocket` crashes when used called in `OnData` method of a Typed JSON.
2. `Session.CalculatePatchAndPushOnWebSocket` causes the app to respond not wrapped in a JSON namespace when called in `OnData` method of a subtree of a Typed JSON.

Link to the app that demonstrates it fully: XXX

#### With both `CalculatePatchAndPushOnWebSocket` enabled

![2017-03-14 08_49_18-helloworld](https://cloud.githubusercontent.com/assets/566463/23890972/6702f4e4-0894-11e7-9d31-9b01eedb728e.png)

#### With first `CalculatePatchAndPushOnWebSocket` commented out (notice that CurrentPage has no namespace - direct child with `Html`, etc)

![2017-03-14 08_55_35-helloworld](https://cloud.githubusercontent.com/assets/566463/23890959/5d7deb54-0894-11e7-9c1c-63b87e833de9.png)

#### With both `CalculatePatchAndPushOnWebSocket` commented out (notice that CurrentPage has a namespace "HelloWorld", as desired)

![2017-03-14 08_56_20-helloworld](https://cloud.githubusercontent.com/assets/566463/23890947/57df5430-0894-11e7-85d6-8af1573dcbee.png)

**Comment by warpech** [comment]
cc @chrhol

**Comment by chrhol** [comment]
I'm not sure what the issue is here. This method does exactly what the name says, it calculates the patches of the public viewmodel, as the public viewmodel looks at the time the method was called. So that means if it's called before parts of the tree have been blended, this is what happens. 

I cannot think of any way to stop or warn appdevelopers from doing this, other than document the behaviour properly, that really visualise what this push does.

As for point 1, I wouldn't say `Session.CalculatePatchAndPushOnWebSocket` crashes. There simply is no session because the viewmodel you are currently in has not been attached to the full tree yet. This is something the developer has to take care to properly handle.

**Comment by chrhol** [comment]
Maybe we should refactor the method to be async in itself? So (we probably need to change the name of it :smile:) we always schedule a task for it, `session.CalculatePatchAndPushOnWebSocketAsync()`

**Comment by warpech** [comment]
> As for point 1, I wouldn't say `Session.CalculatePatchAndPushOnWebSocket` crashes. There simply is no session because the viewmodel you are currently in has not been attached to the full tree yet. This is something the developer has to take care to properly handle.

The biggest problem for me is the unpredictability of the behaviour. Why does line (1) have `Session == null`, whereas in line (2) `Session == Session.Current` yet it causes no-namespace? Both lines are constructed at the same time, since `Spendings` array is bound to the data property `Spendings`.

I am sure that perhaps line (1) might be correct, but then line (2) behaves obviously wrong.

**Comment by warpech** [comment]
Maybe the correct solution is to make it throw an exception when not called in `Session.ScheduleTask`?

**Comment by chrhol** [comment]
> The biggest problem for me is the unpredictability of the behaviour. Why does line (1) have Session == null, whereas in line (2) Session == Session.Current 

Yes, I can understand that, but this is a different topic, and depends on how you write your code. `OnData` is triggered when you set the dataobject (manually or by a binding).

If you do it manually and use this pattern (for example): 
```csharp 
master.CurrentPage = new SpenderPage() { Data = dataObject };
```
Then the property `Session` will be null when `OnData` is triggered, since its triggered before you attach it to the full viewmodel tree.

If you do like this, `Session` property will not be null:
```csharp 
master.CurrentPage = new SpenderPage();
master.CurrentPage.Data = dataObject;
```

As for when `OnData` is triggered by a binding, then the new viewmodel instance will always be attached to the full viewmodel tree before setting `Data`. Thats why it works in the second case, and not in the first.

**Comment by chrhol** [comment]
> Or make it throw an exception when not called in Session.ScheduleTask?

I'm not sure that will help. If you do a push when you are in the middle of updating or creating the viewmodel, you will still do a push that might return something that you didn't expect. And we cannot really detect that from this place in the code you should not be allowed to call push. After all this code will execute the same way, even if it's from a scheduled task.

And also if we do the whole function async, we move a bit further to have a more unified API (`Db.Transact` which is async and waits for completion, `Db.TransactAsync` that is async but does not wait for completion)

Anyone else have any input on how we can improve this, @alemoi, @dan31, @per-samuelsson.
Maybe we need to look at how it could/(will) work in the next version (Starcounter 3.x) and try to move in that direction?

**Comment by warpech** [comment]
TL;DR right now `CalculatePatchAndPushOnWebSocket` shoots you in foot by default :)

Then I think we should change it to `CalculatePatchAndPushOnWebSocketAsync`, as you have proposed.

**Comment by chrhol** [comment]
> TL;DR right now CalculatePatchAndPushOnWebSocket shoots you in foot by default :)

:trollface: A good summary of the discussion and the behaviour of this function.
